#### Imports

In [87]:
import pandas as pd
from random import randint
pd.set_option('display.max_rows', 5)
pd.set_option('display.max_columns', 500)
pd.set_option('display.width', 10000)

#### Download the latest [Pleiades names csv file](https://atlantides.org/downloads/pleiades/dumps/) and read it.

In [2]:
import os
import pandas as pd

In [73]:
df = pd.read_csv("pleiades-places-latest.csv")

#### Print the first two rows

In [5]:
df.head(2)

authors                                            bbox  \
0  Becker, J., T. Elliott    13.4119837, 42.082885, 13.4119837, 42.082885   
1  Becker, J., T. Elliott  11.6285463, 42.4193742, 11.6285463, 42.4193742   

  connectsWith               created          creators  currentVersion  \
0       413005  2016-11-04T16:36:09Z  jbecker, thomase             1.0   
1       413393  2016-11-04T16:39:09Z  jbecker, thomase             2.0   

                                         description  \
0  The post-Roman settlement at Alba Fucens becam...   
1  A major urban sanctuary at Vulci with a long p...   

                                              extent featureTypes geoContext  \
0  {"type": "Point", "coordinates": [13.4119837, ...   settlement        NaN   
1  {"type": "Point", "coordinates": [11.6285463, ...     temple-2        NaN   

   ...              path    reprLat            reprLatLong   reprLong  \
0  ...  /places/48210385  42.082885   42.082885,13.4119837  13.411984   
1  ...  /places/48210386  42.419374  42.4193742,11.6285463  11.628546   

                                tags timePeriods  \
0                                NaN           M   
1  sanctuary, extant remains, temple        ACHR   

                                  timePeriodsKeys  timePeriodsRange  \
0                             mediaeval-byzantine      640.0,1453.0   
1  archaic,classical,hellenistic-republican,roman      -750.0,300.0   

                    title                               uid  
0         Borgo Medievale  ece5760c4c6d42c1a331aad543c4ecc4  
1  Tempio Grande at Vulci  4e06898f2de74dbc9f3a3bdba6d74ba2  

[2 rows x 26 columns]

#### Use the `describe()` method to show basic statistics (mean, median, std, etc.) of numeric columns

In [7]:
p = pd.read_csv("pleiades-places-latest.csv")
p.describe()

currentVersion            id       maxDate       minDate       reprLat  \
count    40132.000000  4.037400e+04  37648.000000  3.764800e+04  32841.000000   
mean         3.861034  5.969153e+12    669.663780 -1.260405e+03     39.379466   
std          2.766998  1.647313e+14    742.790867  5.189146e+04      7.007582   
min          0.000000  4.040000e+02  -9000.000000 -2.600000e+06    -29.543770   
25%          2.000000  2.563382e+05    300.000000 -3.300000e+02     36.093437   
50%          3.000000  5.507355e+05    640.000000 -3.000000e+01     39.049110   
75%          5.000000  8.767158e+05    640.000000 -3.000000e+01     43.471274   
max         67.000000  6.521790e+15   2100.000000  2.000000e+03     62.500000   

           reprLong  
count  32841.000000  
mean      20.256801  
std       17.341363  
min     -117.221676  
25%        8.396650  
50%       17.153337  
75%       31.218177  
max      111.078300

#### Print the name of the columns

In [14]:
print(df.head(0))

Empty DataFrame
Columns: [authors, bbox, connectsWith, created, creators, currentVersion, description, extent, featureTypes, geoContext, hasConnectionsWith, id, locationPrecision, maxDate, minDate, modified, path, reprLat, reprLatLong, reprLong, tags, timePeriods, timePeriodsKeys, timePeriodsRange, title, uid]
Index: []

[0 rows x 26 columns]


#### Find how many _uniques_ values are in the featureTypes column.

In [15]:
len(df["featureTypes"].value_counts())

1235

In [76]:
df["featureTypes"].nunique()

1235

Result: 1235

#### Find Athens
Hint: It is recorded as "Athenae."

In [32]:
df[(df["title"] == "Athenae")]

authors  \
22109  Traill, J., B. Kiesling, S. Gillies, C. Lee, J...   

                                             bbox connectsWith  \
22109  23.619913, 37.937222, 23.849968, 38.019497       579888   

                    created     creators  currentVersion  \
22109  2018-06-07T20:56:36Z  J.S. Traill            28.0   

                                             description  \
22109  A major Greek city-state and the principal cit...   

                                                  extent featureTypes  \
22109  {"type": "Polygon", "coordinates": [[[23.64460...   settlement   

          geoContext  ...            path    reprLat  \
22109  Athina/Athens  ...  /places/579885  37.972454   

                     reprLatLong   reprLong  \
22109  37.972453603,23.726268222  23.726268   

                                                    tags timePeriods  \
22109  dare:ancient=1, dare:major=1, dare:feature=maj...      ACHRLM   

                                         timePeriodsKeys  timePeriodsRange  \
22109  archaic,classical,hellenistic-republican,roman...     -750.0,2100.0   

         title                               uid  
22109  Athenae  aef8271177d5afba233b2b97dcb52af4  

[1 rows x 26 columns]

#### Find all sanctuaries.
Hint: Have a look first in column "featureTypes". There are more than one piece of information in some cells. How does this affect our code? We have to search for cells with strings that contain the string we are looking for.

In [79]:
#df["featureTypes"].str.contains("sanctuary") #filter
df[df["featureTypes"].str.contains("sanctuary", na=False)] #na=False => do not look in na's => apply if error 

authors  \
19     Spann, P., DARMC, R. Talbert, R. Warner, J. Be...   
1006   Fentress, E., DARMC, R. Talbert, R. Warner, J....   
1017                              Elliott, T., J. Becker   
1039                                          Becker, J.   
1083                           Diffendale, D., J. Becker   
...                                                  ...   
39834                     Vitale, V., J. Prag, J. Becker   
39858                                         Becker, J.   
39859                                         Becker, J.   
39866  Walter, D., E. Brod, Y. Yilmaz, J. Becker, T. ...   
39885                                         Becker, J.   

                                                 bbox connectsWith  \
19         -1.143263, 38.208848, -1.143263, 38.208848          NaN   
1006         6.451106, 36.328033, 6.451106, 36.328033          NaN   
1017       63.177708, 30.290889, 63.177708, 30.290889          NaN   
1039     6.6683386, 46.7353535, 6.6683386, 46.7353535          NaN   
1083      22.741983, 39.3889517, 22.74288, 39.3898099       541044   
...                                               ...          ...   
39834  15.1540248, 37.5827598, 15.1540248, 37.5827598       462067   
39858  12.0013698, 42.2940947, 12.0016706, 42.2942853    573378376   
39859  11.8256075, 42.0463351, 11.8256075, 42.0463351          NaN   
39866    29.1270484, 37.926248, 29.1270484, 37.926248       638860   
39885  -0.1894608, 52.5747722, -0.1894608, 52.5747722          NaN   

                    created                                  creators  \
19     2010-09-24T19:02:53Z                                P.O. Spann   
1006   2010-11-10T18:40:19Z                           E.W.B. Fentress   
1017   2023-07-14T14:35:00Z                                   thomase   
1039   2014-04-13T02:10:34Z                                   jbecker   
1083   2021-12-16T22:51:49Z                                  diffenda   
...                     ...                                       ...   
39834  2015-09-15T14:15:47Z                                   vvitale   
39858  2018-03-08T15:29:26Z                                   jbecker   
39859  2018-03-09T02:29:33Z                                   jbecker   
39866  2018-03-15T17:19:12Z  Douglas Walter, Emily Brod, Yeşim Yilmaz   
39885  2013-09-01T00:43:04Z                                   jbecker   

       currentVersion                                        description  \
19                5.0      An ancient place, cited: BAtlas 27 D3 Fortuna   
1006              5.0     An ancient place, cited: BAtlas 31 E4 Chettaba   
1017              3.0  This "Mesopotamian-style temple" site located ...   
1039              3.0                      Site of a Gallo-Roman temple.   
1083              2.0  Pherai was the main cult center of the goddess...   
...               ...                                                ...   
39834             2.0  An archaeological area in Sicily, near Acireal...   
39858             3.0  A rock-cut altar dating to the first half of t...   
39859             6.0  A coastal Etruscan sanctuary that dates ca. 54...   
39866            20.0  The Ploutonion at Hierapolis is a walled sanct...   
39885             2.0  Flag Fen is a Bronze Age ritual site near Pete...   

                                                  extent  \
19     {"type": "Point", "coordinates": [-1.143263, 3...   
1006   {"type": "Point", "coordinates": [6.451106, 36...   
1017   {"type": "Point", "coordinates": [63.177708, 3...   
1039   {"type": "Point", "coordinates": [6.6683386, 4...   
1083   {"type": "Polygon", "coordinates": [[[22.74267...   
...                                                  ...   
39834  {"type": "Point", "coordinates": [15.1540248, ...   
39858  {"type": "Polygon", "coordinates": [[[12.00143...   
39859  {"type": "Point", "coordinates": [11.8256075, ...   
39866  {"type": "Point", "coordinates": [29.1270484, ...   
39885  {"type": "Point", "coordinates

You may get an error 'ValueError: Cannot mask with non-boolean array containing NA / NaN values'. If yes, then use the argument `na=False` within the `str.contains()` function.

Result: 606 rows × 26 columns

#### Find all temples and sanctuaries
Hint: Have a look first in column "featureTypes". There are more than one piece of information in some cells. How does this affect your code?

In [39]:
df[(df["featureTypes"] == "sanctuary") | (df["featureTypes"] == "temple")]

authors  \
19     Spann, P., DARMC, R. Talbert, R. Warner, J. Be...   
1039                                          Becker, J.   
1137                                           Mckee, G.   
1334   Loseby, S., DARMC, R. Talbert, S. Gillies, J. ...   
1381   Loseby, S., DARMC, R. Talbert, S. Gillies, J. ...   
...                                                  ...   
39793                         Rosalinsky, Z., T. Elliott   
39858                                         Becker, J.   
39885                                         Becker, J.   
39937                      DARMC, S. Gillies, T. Elliott   
40310                      DARMC, S. Gillies, T. Elliott   

                                                 bbox connectsWith  \
19         -1.143263, 38.208848, -1.143263, 38.208848          NaN   
1039     6.6683386, 46.7353535, 6.6683386, 46.7353535          NaN   
1137   24.3947611, 37.4084972, 24.3949441, 37.4086349       570402   
1334         4.232932, 44.199232, 4.232932, 44.199232          NaN   
1381           5.681783, 44.057257, 5.712445, 44.0619          NaN   
...                                               ...          ...   
39793  33.4590083, 29.0364787, 33.4598748, 29.0368794       746811   
39858  12.0013698, 42.2940947, 12.0016706, 42.2942853    573378376   
39885  -0.1894608, 52.5747722, -0.1894608, 52.5747722          NaN   
39937      16.749739, 31.146289, 16.749739, 31.146289          NaN   
40310      14.461641, 32.477054, 14.461641, 32.477054          NaN   

                    created     creators  currentVersion  \
19     2010-09-24T19:02:53Z   P.O. Spann             5.0   
1039   2014-04-13T02:10:34Z      jbecker             3.0   
1137   2023-04-18T15:40:39Z       gmckee             4.0   
1334   2009-11-01T16:50:13Z    S. Loseby             7.0   
1381   2009-11-01T16:51:06Z    S. Loseby             7.0   
...                     ...          ...             ...   
39793  2022-06-29T17:58:53Z  zrosalinsky             2.0   
39858  2018-03-08T15:29:26Z      jbecker             3.0   
39885  2013-09-01T00:43:04Z      jbecker             2.0   
39937  2011-09-09T21:20:28Z        DARMC             2.0   
40310  2011-09-09T21:31:31Z        DARMC             2.0   

                                             description  \
19         An ancient place, cited: BAtlas 27 D3 Fortuna   
1039                       Site of a Gallo-Roman temple.   
1137   Religious structure in Kythnos constructed in ...   
1334   The site of sulphur springs that also revealed...   
1381             A Celto-Ligurian oppidum and sanctuary.   
...                                                  ...   
39793  Sanctuary of Hathor built in the Middle Kingdo...   
39858  A rock-cut altar dating to the first half of t...   
39885  Flag Fen is a Bronze Age ritual site near Pete...   
39937  An unnamed monument, settlement, or villa of S...   
40310                    An unnamed place of Tripolitana   

                                                  extent featureTypes  \
19     {"type": "Point", "coordinates": [-1.143263, 3...    sanctuary   
1039   {"type": "Point", "coordinates": [6.6683386, 4...    sanctuary   
1137   {"type": "Polygon", "coordinates": [[[24.39482...    sanctuary   
1334   {"type": "Point", "coordinates": [4.232932, 44...    sanctuary   
1381   {"type": "LineString", "coordinates": [[5.7124...    sanctuary   
...                                                  ...          ...   
39793  {"type": "Polygon", "coordinates": [[[33.45927...    sanctuary   
39858  {"type": "Polygon", "coordinates": [[[12.00143...    sanctuary   
39885  {"type": "Point", "coordinates": [-0.1894608, ...    sanctuary   
39937  {"type": "Point", "coordinates": [16.749739, 3...       temple   
40310  {"type": "Point", "coordinates": [14.461641, 3...       temple   

                            geoContext  ...               path    reprLat  \
19      Cueva Negra, 13 miles N Murcia  ...     /places/265899  38.208848   
1039          

#### Find which is the northernest, southernest, easternest, and westernest sanctuary of the database.

##### northernest

In [93]:
df1 = df[(df["featureTypes"] == "sanctuary")]
df1[(df1["reprLat"] == df1["reprLat"].max())] #using a different dataframe df1 to find specific columns 
#sorting or using maximum value => like the Athens 

authors                                            bbox connectsWith               created creators  currentVersion                                        description                                             extent featureTypes geoContext hasConnectionsWith         id locationPrecision  maxDate  minDate              modified               path    reprLat            reprLatLong  reprLong            tags timePeriods     timePeriodsKeys timePeriodsRange     title                               uid
39885  Becker, J.  -0.1894608, 52.5747722, -0.1894608, 52.5747722          NaN  2013-09-01T00:43:04Z  jbecker             2.0  Flag Fen is a Bronze Age ritual site near Pete...  {"type": "Point", "coordinates": [-0.1894608, ...    sanctuary        NaN                NaN  535275912           precise   -800.0  -2500.0  2021-02-02T23:38:25Z  /places/535275912  52.574772  52.5747722,-0.1894608 -0.189461  extant remains           B  bronze-age-britain   -2500.0,-800.0  Flag Fen  70327170197c39f92804c0a33f3afe76

##### southernest

In [94]:
df1 = df[(df["featureTypes"] == "sanctuary")]
df1[(df1["reprLat"] == df1["reprLat"].min())]

authors                                      bbox connectsWith               created     creators  currentVersion                                        description                                             extent featureTypes                                         geoContext hasConnectionsWith     id locationPrecision  maxDate  minDate              modified           path   reprLat         reprLatLong   reprLong                                               tags timePeriods     timePeriodsKeys timePeriodsRange       title                               uid
27278  Erdosy, M., DARMC, R. Talbert, S. Gillies, T. ...  80.497715, 6.808048, 80.497715, 6.808048        50298  2010-09-23T01:01:28Z  M.U. Erdosy             4.0  An ancient place, cited: BAtlas 5 inset Oulipp...  {"type": "Point", "coordinates": [80.497715, 6...    sanctuary  town, or possibly “sacred footprint” on Adam’s...                NaN  50184           precise    640.0    -30.0  2016-09-10T02:48:12Z  /places/50184  6.808048  6.808048,80.497715  80.497715  dare:ancient=1, dare:major=0, dare:feature=temple          RL  roman,late-antique      -30.0,640.0  Oulippada?  eaeb6d1cc037a3a417141c1749adc300

##### easternest

In [95]:
df1 = df[(df["featureTypes"] == "sanctuary")]
df1[(df1["reprLong"] == df1["reprLong"].max())]

authors                                      bbox connectsWith               created     creators  currentVersion                                        description                                             extent featureTypes                                         geoContext hasConnectionsWith     id locationPrecision  maxDate  minDate              modified           path   reprLat         reprLatLong   reprLong                                               tags timePeriods     timePeriodsKeys timePeriodsRange       title                               uid
27278  Erdosy, M., DARMC, R. Talbert, S. Gillies, T. ...  80.497715, 6.808048, 80.497715, 6.808048        50298  2010-09-23T01:01:28Z  M.U. Erdosy             4.0  An ancient place, cited: BAtlas 5 inset Oulipp...  {"type": "Point", "coordinates": [80.497715, 6...    sanctuary  town, or possibly “sacred footprint” on Adam’s...                NaN  50184           precise    640.0    -30.0  2016-09-10T02:48:12Z  /places/50184  6.808048  6.808048,80.497715  80.497715  dare:ancient=1, dare:major=0, dare:feature=temple          RL  roman,late-antique      -30.0,640.0  Oulippada?  eaeb6d1cc037a3a417141c1749adc300

##### westernest

In [96]:
df1 = df[(df["featureTypes"] == "sanctuary")]
df1[(df1["reprLong"] == df1["reprLong"].min())]

authors                                          bbox connectsWith               created                       creators  currentVersion                                        description                                             extent featureTypes                                 geoContext hasConnectionsWith         id locationPrecision  maxDate  minDate              modified               path    reprLat           reprLatLong  reprLong                                               tags timePeriods     timePeriodsKeys timePeriodsRange           title                               uid
15822                                         Becker, J.  -9.472854, 38.8242125, -9.472854, 38.8242125          NaN  2013-11-12T23:33:37Z                        jbecker             1.0  A Roman sanctuary located at the mouth of the ...  {"type": "Point", "coordinates": [-9.472854, 3...    sanctuary                                        NaN                NaN  614803679           precise    640.0    -30.0  2018-10-05T11:06:06Z  /places/614803679  38.824213  38.8242125,-9.472854 -9.472854                                                NaN          RL  roman,late-antique      -30.0,640.0   Alto da Vigia  3faa112b48de1f5a4c4abdd8c044e4b8
22968  Jr., F. Stanley, R. Knapp, DARMC, R. Talbert, ...  -9.472854, 38.8242125, -9.472854, 38.8242125          NaN  2021-11-13T02:43:18Z  Jr., F.H. Stanley, R.C. Knapp             7.0  An ancient place, cited: BAtlas 26 A3 Foz de C...  {"type": "Point", "coordinates": [-9.472854, 3...    sanctuary  Foz de Colares, sanctuary near Sintra POR                NaN     256172           precise    300.0    -30.0  2022-07-14T13:03:42Z     /places/256172  38.824213  38.8242125,-9.472854 -9.472854  dare:major=0, dare:ancient=0, dare:feature=temple           R               roman      -30.0,300.0  Foz de Colares  ca1d5692afbaf70b305b438035da787c

---

#### By using the methin `isin()`, check which items of list1 are _not_ in list2.
Hint: The following code creates two list of numbers from 1 to 100, removes randomly one number from each of them, and transforms the lists to Series.

In [88]:
list1 = [i for i in range(1,101)]
list1.pop(randint(1, 101))
ser1 = pd.Series(list1)
list2 = [i for i in range(1,101)]
list2.pop(randint(1, 101))
ser2 = pd.Series(list2)

In [89]:
ser2[~ser2.isin(ser1)]

28    29
dtype: int64

---